File: AppBox/Common/DefaultAppHandler/DefaultAppHandler.m
//
//  DefaultAppHandler.m
//  AppBox
//
//  Created by Vineet Choudhary on 07/06/26.
//  Copyright © 2026 Developer Insider. All rights reserved.
//

#import "DefaultAppHandler.h"

NSNotificationName const ABDefaultIPAHandlerDidChangeNotification = @"ABDefaultIPAHandlerDidChangeNotification";

@implementation DefaultAppHandler

+ (BOOL)isDefaultIPAHandler {
    NSString *bundleId = [[NSBundle mainBundle] bundleIdentifier];
    CFStringRef defaultHandler = LSCopyDefaultRoleHandlerForContentType(CFSTR("com.apple.itunes.ipa"), kLSRolesAll);
    if (defaultHandler == NULL) return NO;
    BOOL isDefault = [(__bridge_transfer NSString *)defaultHandler isEqualToString:bundleId];
    return isDefault;
}

+ (void)setAsDefaultIPAHandler {
    NSString *bundleId = [[NSBundle mainBundle] bundleIdentifier];
    BOOL wasPreviouslyDefault = [self isDefaultIPAHandler];
    
    OSStatus status = LSSetDefaultRoleHandlerForContentType(CFSTR("com.apple.itunes.ipa"), kLSRolesAll, (__bridge CFStringRef)bundleId);
    if (status != noErr) {
        DDLogError(@"Failed to set AppBox as default IPA handler (error %d).", (int)status);
    } else {
        DDLogInfo(@"AppBox set as default application for .ipa files.");
    }
    
    // Observe app becoming active to detect when system popup closes
    [self observeAppActivationWithPreviousState:wasPreviouslyDefault];
}

+ (void)removeAsDefaultIPAHandler {
    NSString *appBoxBundleId = [[NSBundle mainBundle] bundleIdentifier];
    BOOL wasPreviouslyDefault = [self isDefaultIPAHandler];
    
    // Find all handlers for IPA content type and pick the first one that isn't AppBox
    CFArrayRef allHandlers = LSCopyAllRoleHandlersForContentType(CFSTR("com.apple.itunes.ipa"), kLSRolesAll);
    if (allHandlers != NULL) {
        NSArray *handlers = (__bridge_transfer NSArray *)allHandlers;
        for (NSString *handler in handlers) {
            if (![handler isEqualToString:appBoxBundleId]) {
                OSStatus status = LSSetDefaultRoleHandlerForContentType(CFSTR("com.apple.itunes.ipa"), kLSRolesAll, (__bridge CFStringRef)handler);
                if (status == noErr) {
                    DDLogInfo(@"Default IPA handler change requested to: %@", handler);
                    break;
                }
            }
        }
    } else {
        // Fallback: set Archive Utility as default since IPA is a ZIP archive
        OSStatus status = LSSetDefaultRoleHandlerForContentType(CFSTR("com.apple.itunes.ipa"), kLSRolesAll, CFSTR("com.apple.archiveutility"));
        if (status == noErr) {
            DDLogInfo(@"Default IPA handler change requested to Archive Utility.");
        } else {
            DDLogError(@"Failed to remove AppBox as default IPA handler (error %d).", (int)status);
        }
    }
    
    // Observe app becoming active to detect when system popup closes
    [self observeAppActivationWithPreviousState:wasPreviouslyDefault];
}

#pragma mark - Private

+ (void)observeAppActivationWithPreviousState:(BOOL)wasPreviouslyDefault {
    // Wait 2 seconds for system popup to appear, then poll every 0.5s for state change
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        // Post the current real status immediately
        BOOL isNowDefault = [self isDefaultIPAHandler];
        [[NSNotificationCenter defaultCenter]
            postNotificationName:ABDefaultIPAHandlerDidChangeNotification
            object:nil
            userInfo:@{@"isDefault": @(isNowDefault)}];
        
        // Then keep polling every 0.5s until state changes or timeout (30s)
        __block NSInteger attempts = 0;
        __block BOOL lastKnownState = isNowDefault;
        NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:0.5 repeats:YES block:^(NSTimer *t) {
            attempts++;
            BOOL currentState = [self isDefaultIPAHandler];
            
            if (currentState != lastKnownState) {
                lastKnownState = currentState;
                [[NSNotificationCenter defaultCenter]
                    postNotificationName:ABDefaultIPAHandlerDidChangeNotification
                    object:nil
                    userInfo:@{@"isDefault": @(currentState)}];
            }
            
            // Stop after state changed from original or timeout
            if (currentState != wasPreviouslyDefault || attempts >= 60) {
                [t invalidate];
                DDLogInfo(@"IPA handler state resolved: was=%d, now=%d (attempts=%ld)", wasPreviouslyDefault, currentState, (long)attempts);
            }
        }];
        [[NSRunLoop mainRunLoop] addTimer:timer forMode:NSRunLoopCommonModes];
    });
}

@end

File: AppBox/Common/DefaultAppHandler/DefaultAppHandler.swift
//
//  DefaultAppHandler.swift
//  AppBox

import Foundation
import CoreServices
import os

public final class DefaultAppHandler: NSObject {

    /// Posted (on the main thread) when the default-.ipa-handler state may have changed.
    public static let didChangeNotificationName = "ABDefaultIPAHandlerDidChangeNotification"

    private static let log = Logger(subsystem: "com.developerinsider.AppBox", category: "DefaultAppHandler")
    private static let ipaContentType = "com.apple.itunes.ipa" as CFString

    public class func isDefaultIPAHandler() -> Bool {
        guard let bundleId = Bundle.main.bundleIdentifier else { return false }
        guard let handler = LSCopyDefaultRoleHandlerForContentType(ipaContentType, .all)?.takeRetainedValue() else {
            return false
        }
        return (handler as String) == bundleId
    }

    public class func setAsDefaultIPAHandler() {
        guard let bundleId = Bundle.main.bundleIdentifier else { return }
        let wasPreviouslyDefault = isDefaultIPAHandler()
        let status = LSSetDefaultRoleHandlerForContentType(ipaContentType, .all, bundleId as CFString)
        if status != noErr {
            log.error("Failed to set AppBox as default IPA handler (error \(status, privacy: .public)).")
        } else {
            log.info("AppBox set as default application for .ipa files.")
        }
        observeAppActivation(wasPreviouslyDefault: wasPreviouslyDefault)
    }

    public class func removeAsDefaultIPAHandler() {
        guard let appBoxBundleId = Bundle.main.bundleIdentifier else { return }
        let wasPreviouslyDefault = isDefaultIPAHandler()

        if let handlers = LSCopyAllRoleHandlersForContentType(ipaContentType, .all)?.takeRetainedValue() as? [String] {
            for handler in handlers where handler != appBoxBundleId {
                if LSSetDefaultRoleHandlerForContentType(ipaContentType, .all, handler as CFString) == noErr {
                    log.info("Default IPA handler change requested to: \(handler, privacy: .public)")
                    break
                }
            }
        } else {
            if LSSetDefaultRoleHandlerForContentType(ipaContentType, .all, "com.apple.archiveutility" as CFString) == noErr {
                log.info("Default IPA handler change requested to Archive Utility.")
            } else {
                log.error("Failed to remove AppBox as default IPA handler.")
            }
        }
        observeAppActivation(wasPreviouslyDefault: wasPreviouslyDefault)
    }

    // MARK: - Private

    /// After a change, macOS shows a confirmation popup we can't observe directly, so we wait then poll the real state (a 2s delay, then every 0.5s up to ~30s).
    private static func observeAppActivation(wasPreviouslyDefault: Bool) {
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            let isNowDefault = isDefaultIPAHandler()
            post(isDefault: isNowDefault)

            var attempts = 0
            var lastKnownState = isNowDefault
            let timer = Timer(timeInterval: 0.5, repeats: true) { t in
                attempts += 1
                let currentState = isDefaultIPAHandler()
                if currentState != lastKnownState {
                    lastKnownState = currentState
                    post(isDefault: currentState)
                }
                if currentState != wasPreviouslyDefault || attempts >= 60 {
                    t.invalidate()
                    log.info("IPA handler state resolved: was=\(wasPreviouslyDefault, privacy: .public), now=\(currentState, privacy: .public) (attempts=\(attempts, privacy: .public))")
                }
            }
            RunLoop.main.add(timer, forMode: .common)
        }
    }

    private static func post(isDefault: Bool) {
        NotificationCenter.default.post(
			name: Notification.Name(didChangeNotificationName),
			object: nil,
			userInfo: ["isDefault": isDefault])
    }
}
